File: desktop/src/agent/conversationPrivacy.ts
import type { UIMessage } from "ai";
import type { AgentToolContext } from "./agentContext";
import type { ConversationRecord } from "./chatStore";

export const CONVERSATION_PRIVACY_VERSION = 1;

const REDACTED = "[REDACTED]";
const LOCAL_PATH = "[LOCAL_PATH]";

const SENSITIVE_KEYS = [
  "password",
  "passwd",
  "passphrase",
  "secret",
  "token",
  "accesstoken",
  "refreshtoken",
  "apikey",
  "authorization",
  "auth",
  "credential",
  "privatekey",
  "webhook",
  "cookie",
  "databaseurl",
  "connectionstring",
];

const LARGE_PRIVATE_KEYS = new Set([
  "raw",
  "body",
  "logtail",
  "manifest",
  "downloadurl",
  "sha1",
  "sha256",
  "sha512",
  "sessionid",
  "outputpath",
]);

const WIKI_TOOL_TYPES = new Set(["tool-wiki_search", "tool-wiki_open"]);

export interface PublicShareTextPart {
  type: "text";
  text: string;
}

export interface PublicShareQuestionPart {
  type: "tool-ask_user_question";
  input: {
    question: string;
    options: Array<{ label: string; description?: string }>;
  };
  output: { selected: string[] };
}

export interface PublicShareMessage {
  role: "user" | "assistant";
  parts: Array<PublicShareTextPart | PublicShareQuestionPart>;
}

function normalizedKey(key: string): string {
  return key.toLowerCase().replace(/[^a-z0-9]/g, "");
}

function isSensitiveKey(key: string): boolean {
  const normalized = normalizedKey(key);
  return SENSITIVE_KEYS.some((part) => normalized === part || normalized.endsWith(part));
}

function privateContextValues(context: AgentToolContext | null | undefined): string[] {
  const values = [
    context?.root,
    context?.instance?.root,
    ...(context?.instance?.sourcePaths ?? []),
    context?.wiki?.root,
    ...(context?.wiki?.sourcePaths ?? []),
  ].filter((value): value is string => typeof value === "string" && value.length > 0);
  const variants = new Set<string>();
  for (const value of values) {
    variants.add(value);
    variants.add(value.replaceAll("\\", "/"));
    variants.add(value.replaceAll("/", "\\"));
  }
  return [...variants].sort((left, right) => right.length - left.length);
}

function replacePrivateValues(value: string, privateValues: string[]): string {
  let sanitized = value;
  for (const privateValue of privateValues) {
    sanitized = sanitized.split(privateValue).join(LOCAL_PATH);
  }
  return sanitized;
}

function sanitizeCredentialText(value: string): string {
  return value
    .replace(/\bBearer\s+[A-Za-z0-9._~+\/-]+=*/gi, `Bearer ${REDACTED}`)
    .replace(/\b(?:sk|pk|ghp|glpat)-[A-Za-z0-9_-]{12,}\b/g, REDACTED)
    .replace(/\bCookie\s*:\s*[^\r\n]+/gi, `Cookie: ${REDACTED}`)
    .replace(
      /(^|[\s,{;])(password|passwd|passphrase|secret|token|access[_-]?token|refresh[_-]?token|api[_-]?key|authorization|auth|credential|private[_-]?key|webhook|cookie|database[_-]?url|connection[_-]?string)(\s*[:=]\s*)("[^"]*"|'[^']*'|[^\r\n,;}]+)/gim,
      (_match, prefix: string, key: string, separator: string) =>
        `${prefix}${key}${separator}${REDACTED}`,
    );
}

function sanitizeLocalPathText(value: string): string {
  return value
    .replace(/file:\/\/[^\r\n"'`<>()]+/gi, LOCAL_PATH)
    .replace(/\\\\[^\r\n"'`<>()]+/g, LOCAL_PATH)
    .replace(/~[\\/][^\r\n"'`<>()]+/g, LOCAL_PATH)
    .replace(/[A-Za-z]:[\\/][^\r\n"'`<>()]+/g, LOCAL_PATH)
    .replace(/\/(?:Users|home|var|tmp|private|Volumes|opt|etc|Library|Applications)\/[^\r\n"'`<>()]+/g, LOCAL_PATH)
    .replace(/(^|[\s("'`])\/(?!\/)(?:[^/\s"'`<>()]+\/)+[^\s"'`<>()]+/g, (_match, prefix: string) => `${prefix}${LOCAL_PATH}`);
}

function sanitizeString(value: string, privateValues: string[]): string {
  return sanitizeCredentialText(
    sanitizeLocalPathText(replacePrivateValues(value, privateValues)),
  );
}

function sanitizePublicString(value: string): string {
  return sanitizeCredentialText(sanitizeLocalPathText(value))
    .replace(/https?:\/\/[^\s"'`<>()]+/gi, REDACTED)
    .replace(/\b[a-f0-9]{40,128}\b/gi, REDACTED);
}

function sanitizeValue(
  value: unknown,
  key: string,
  depth: number,
  sanitizeText: (text: string) => string,
): unknown {
  if (depth > 24) return REDACTED;
  if (isSensitiveKey(key) || LARGE_PRIVATE_KEYS.has(normalizedKey(key))) return REDACTED;
  if (typeof value === "string") return sanitizeText(value);
  if (Array.isArray(value)) {
    return value.map((item) => sanitizeValue(item, key, depth + 1, sanitizeText));
  }
  if (!value || typeof value !== "object") return value;

  const result: Record<string, unknown> = {};
  for (const [entryKey, entryValue] of Object.entries(value as Record<string, unknown>)) {
    if (entryKey === "providerMetadata") continue;
    result[entryKey] = sanitizeValue(entryValue, entryKey, depth + 1, sanitizeText);
  }
  return result;
}

function projectMessageForPersistence(message: UIMessage, privateValues: string[]): UIMessage {
  const sanitizeText = (value: string) => sanitizeString(value, privateValues);
  const raw = message as unknown as Record<string, unknown>;
  const rawParts = Array.isArray(raw.parts) ? raw.parts : [];
  const parts = rawParts.flatMap((part) => {
    if (!part || typeof part !== "object") return [part];
    const toolPart = part as Record<string, unknown>;
    if (toolPart.type === "reasoning") return [];
    if (typeof toolPart.type === "string" && WIKI_TOOL_TYPES.has(toolPart.type)) {
      const withoutPrivateOutput = "output" in toolPart
        ? {
            ...toolPart,
            output: {
              privacyRedacted: true,
              reason: "instance_content_not_persisted",
            },
          }
        : toolPart;
      return [sanitizeValue(withoutPrivateOutput, "part", 0, sanitizeText)];
    }
    return [sanitizeValue(toolPart, "part", 0, sanitizeText)];
  });
  const messageFields = { ...raw };
  delete messageFields.parts;
  const projected = sanitizeValue(
    messageFields,
    "message",
    0,
    sanitizeText,
  ) as Record<string, unknown>;
  return { ...projected, parts } as unknown as UIMessage;
}

function projectToolContext(context: AgentToolContext | null | undefined): AgentToolContext | null {
  if (!context) return null;
  const projected: AgentToolContext = {};
  if (context.mode) projected.mode = context.mode;
  if (context.instance) {
    projected.instance = {
      modpackId: context.instance.modpackId,
      instanceId: context.instance.instanceId,
      mcVersion: context.instance.mcVersion,
      loader: context.instance.loader,
    };
  } else if (context.wiki) {
    projected.wiki = {
      modpackId: context.wiki.modpackId,
      instanceId: context.wiki.instanceId,
    };
  }
  return projected;
}

export function projectConversationForPersistence(record: ConversationRecord): ConversationRecord {
  const privateValues = privateContextValues(record.toolContext);
  return {
    id: record.id,
    createdAt: record.createdAt,
    updatedAt: record.updatedAt,
    title: sanitizeString(record.title, privateValues),
    messages: record.messages.map((message) => projectMessageForPersistence(message, privateValues)),
    toolContext: projectToolContext(record.toolContext),
    ownerId: record.ownerId ?? null,
    privacyVersion: CONVERSATION_PRIVACY_VERSION,
  };
}

function publicQuestionPart(raw: Record<string, unknown>): PublicShareQuestionPart | null {
  if (!raw.input || typeof raw.input !== "object") return null;
  const input = raw.input as Record<string, unknown>;
  if (typeof input.question !== "string") return null;
  const options = Array.isArray(input.options)
    ? input.options.flatMap((option) => {
        if (!option || typeof option !== "object") return [];
        const value = option as Record<string, unknown>;
        if (typeof value.label !== "string") return [];
        return [{
          label: sanitizePublicString(value.label),
          ...(typeof value.description === "string"
            ? { description: sanitizePublicString(value.description) }
            : {}),
        }];
      })
    : [];
  const output = raw.output && typeof raw.output === "object"
    ? raw.output as Record<string, unknown>
    : {};
  const selected = Array.isArray(output.selected)
    ? output.selected.filter((value): value is string => typeof value === "string")
        .map(sanitizePublicString)
    : [];
  return {
    type: "tool-ask_user_question",
    input: {
      question: sanitizePublicString(input.question),
      options,
    },
    output: { selected },
  };
}

function visibleParts(message: UIMessage): Array<PublicShareTextPart | PublicShareQuestionPart> {
  const parts = (message as unknown as { parts?: unknown[] }).parts ?? [];
  const visible: Array<PublicShareTextPart | PublicShareQuestionPart> = [];
  for (const part of parts) {
    if (!part || typeof part !== "object") continue;
    const raw = part as Record<string, unknown>;
    if (raw.type === "text" && typeof raw.text === "string") {
      visible.push({ type: "text", text: sanitizePublicString(raw.text) });
      continue;
    }
    if (raw.type !== "tool-ask_user_question") continue;
    const question = publicQuestionPart(raw);
    if (question) visible.push(question);
  }
  return visible;
}

export function projectMessagesForPublicShare(messages: UIMessage[]): PublicShareMessage[] {
  const projected: PublicShareMessage[] = [];
  for (const message of messages) {
    if (message.role !== "user" && message.role !== "assistant") continue;
    const parts = visibleParts(message);
    if (parts.length === 0) continue;
    projected.push({ role: message.role, parts });
  }
  return projected;
}

File: crates/mc-core/src/agent/conversation_privacy.rs
use serde_json::{Map, Value};

use crate::error::{CoreError, Result};

use super::tools::sanitize_private_text;

pub const CONVERSATION_PRIVACY_VERSION: u64 = 1;

pub fn sanitize_conversation_text(text: &str) -> String {
    sanitize_persisted_text(text)
}

const REDACTED: &str = "[REDACTED]";
const WIKI_TOOL_TYPES: [&str; 2] = ["tool-wiki_search", "tool-wiki_open"];
const LARGE_PRIVATE_KEYS: [&str; 9] = [
    "raw",
    "body",
    "logtail",
    "manifest",
    "downloadurl",
    "sha1",
    "sha256",
    "sha512",
    "sessionid",
];

pub fn project_conversation_record(record: &Value, owner_id: Option<&str>) -> Result<Value> {
    let object = record
        .as_object()
        .ok_or_else(|| CoreError::other("conversation record must be an object"))?;
    let id = object
        .get("id")
        .and_then(Value::as_str)
        .filter(|id| !id.is_empty())
        .ok_or_else(|| CoreError::other("conversation record requires id"))?;
    let messages = object
        .get("messages")
        .and_then(Value::as_array)
        .ok_or_else(|| CoreError::other("conversation record requires messages"))?;

    let mut projected = Map::new();
    projected.insert("id".into(), id.into());
    for key in ["createdAt", "updatedAt"] {
        if let Some(value) = object.get(key) {
            projected.insert(key.into(), value.clone());
        }
    }
    projected.insert(
        "title".into(),
        object
            .get("title")
            .and_then(Value::as_str)
            .map(sanitize_persisted_text)
            .unwrap_or_default()
            .into(),
    );
    projected.insert(
        "messages".into(),
        messages
            .iter()
            .filter_map(project_message)
            .collect::<Vec<_>>()
            .into(),
    );
    projected.insert(
        "toolContext".into(),
        object
            .get("toolContext")
            .and_then(project_tool_context)
            .unwrap_or(Value::Null),
    );
    projected.insert(
        "ownerId".into(),
        owner_id.map(Value::from).unwrap_or(Value::Null),
    );
    projected.insert("privacyVersion".into(), CONVERSATION_PRIVACY_VERSION.into());
    Ok(Value::Object(projected))
}

pub fn project_public_share(payload: &Value) -> Result<Value> {
    let messages = payload
        .get("messages")
        .and_then(Value::as_array)
        .ok_or_else(|| CoreError::other("public share payload requires messages"))?;
    let messages = messages
        .iter()
        .filter_map(|message| {
            let role = message.get("role")?.as_str()?;
            if role != "user" && role != "assistant" {
                return None;
            }
            let parts = message
                .get("parts")?
                .as_array()?
                .iter()
                .filter_map(project_public_part)
                .collect::<Vec<_>>();
            (!parts.is_empty()).then(|| {
                serde_json::json!({
                    "role": role,
                    "parts": parts,
                })
            })
        })
        .collect::<Vec<_>>();
    Ok(serde_json::json!({ "messages": messages }))
}

pub fn conversation_record_owner(record: &Value) -> Option<&str> {
    record.get("ownerId").and_then(Value::as_str)
}

pub fn conversation_record_visible_to(record: &Value, owner_id: Option<&str>) -> bool {
    match (conversation_record_owner(record), owner_id) {
        (None, _) => true,
        (Some(record_owner), Some(owner)) => record_owner == owner,
        (Some(_), None) => false,
    }
}

fn project_message(message: &Value) -> Option<Value> {
    let object = message.as_object()?;
    let mut projected = Map::new();
    for (key, value) in object {
        if key == "parts" || key == "providerMetadata" {
            continue;
        }
        insert_sanitized_entry(
            &mut projected,
            key,
            sanitize_value(value, key, 0, false),
            false,
        );
    }
    let parts = object
        .get("parts")
        .and_then(Value::as_array)
        .into_iter()
        .flatten()
        .filter_map(project_part)
        .collect::<Vec<_>>();
    projected.insert("parts".into(), parts.into());
    Some(Value::Object(projected))
}

fn project_part(part: &Value) -> Option<Value> {
    let object = part.as_object()?;
    let part_type = object.get("type").and_then(Value::as_str);
    if part_type == Some("reasoning") {
        return None;
    }
    let mut projected = sanitize_value(part, "part", 0, false);
    if part_type.is_some_and(|kind| WIKI_TOOL_TYPES.contains(&kind)) {
        if let Some(object) = projected.as_object_mut() {
            if object.contains_key("output") {
                object.insert(
                    "output".into(),
                    serde_json::json!({
                        "privacyRedacted": true,
                        "reason": "instance_content_not_persisted",
                    }),
                );
            }
        }
    }
    Some(projected)
}

fn project_tool_context(context: &Value) -> Option<Value> {
    let object = context.as_object()?;
    let mut projected = Map::new();
    copy_string(object, &mut projected, "mode");
    if let Some(instance) = object.get("instance").and_then(Value::as_object) {
        projected.insert(
            "instance".into(),
            select_strings(
                instance,
                &["modpackId", "instanceId", "mcVersion", "loader"],
            ),
        );
    } else if let Some(wiki) = object.get("wiki").and_then(Value::as_object) {
        projected.insert(
            "wiki".into(),
            select_strings(wiki, &["modpackId", "instanceId"]),
        );
    }
    Some(Value::Object(projected))
}

fn select_strings(object: &Map<String, Value>, keys: &[&str]) -> Value {
    let mut selected = Map::new();
    for key in keys {
        copy_string(object, &mut selected, key);
    }
    Value::Object(selected)
}

fn copy_string(source: &Map<String, Value>, target: &mut Map<String, Value>, key: &str) {
    if let Some(value) = source.get(key).and_then(Value::as_str) {
        target.insert(key.into(), sanitize_persisted_text(value).into());
    }
}

fn project_public_part(part: &Value) -> Option<Value> {
    let object = part.as_object()?;
    match object.get("type")?.as_str()? {
        "text" => Some(serde_json::json!({
            "type": "text",
            "text": sanitize_public_text(object.get("text")?.as_str()?),
        })),
        "tool-ask_user_question" => project_public_question(object),
        _ => None,
    }
}

fn project_public_question(part: &Map<String, Value>) -> Option<Value> {
    let input = part.get("input")?.as_object()?;
    let question = sanitize_public_text(input.get("question")?.as_str()?);
    let options = input
        .get("options")
        .and_then(Value::as_array)
        .into_iter()
        .flatten()
        .filter_map(|option| {
            let option = option.as_object()?;
            let mut projected = Map::new();
            projected.insert(
                "label".into(),
                sanitize_public_text(option.get("label")?.as_str()?).into(),
            );
            if let Some(description) = option.get("description").and_then(Value::as_str) {
                projected.insert(
                    "description".into(),
                    sanitize_public_text(description).into(),
                );
            }
            Some(Value::Object(projected))
        })
        .collect::<Vec<_>>();
    let selected = part
        .get("output")
        .and_then(|output| output.get("selected"))
        .and_then(Value::as_array)
        .into_iter()
        .flatten()
        .filter_map(Value::as_str)
        .map(sanitize_public_text)
        .collect::<Vec<_>>();
    Some(serde_json::json!({
        "type": "tool-ask_user_question",
        "input": { "question": question, "options": options },
        "output": { "selected": selected },
    }))
}

fn sanitize_value(value: &Value, key: &str, depth: usize, public: bool) -> Value {
    if depth > 24 || is_sensitive_key(key) || is_large_private_key(key) {
        return REDACTED.into();
    }
    match value {
        Value::String(text) => {
            if public {
                sanitize_public_text(text).into()
            } else {
                sanitize_persisted_text(text).into()
            }
        }
        Value::Array(items) => items
            .iter()
            .map(|item| sanitize_value(item, key, depth + 1, public))
            .collect::<Vec<_>>()
            .into(),
        Value::Object(object) => {
            let mut sanitized = Map::new();
            for (entry_key, entry_value) in object {
                if entry_key == "providerMetadata" {
                    continue;
                }
                insert_sanitized_entry(
                    &mut sanitized,
                    entry_key,
                    sanitize_value(entry_value, entry_key, depth + 1, public),
                    public,
                );
            }
            Value::Object(sanitized)
        }
        _ => value.clone(),
    }
}

fn insert_sanitized_entry(target: &mut Map<String, Value>, key: &str, value: Value, public: bool) {
    let base = if public {
        sanitize_public_text(key)
    } else {
        sanitize_persisted_text(key)
    };
    let mut candidate = base.clone();
    let mut suffix = 2;
    while target.contains_key(&candidate) {
        candidate = format!("{base}#{suffix}");
        suffix += 1;
    }
    target.insert(candidate, value);
}

fn is_sensitive_key(key: &str) -> bool {
    let normalized = normalized_key(key);
    [
        "password",
        "passwd",
        "passphrase",
        "secret",
        "token",
        "accesstoken",
        "refreshtoken",
        "apikey",
        "authorization",
        "auth",
        "credential",
        "privatekey",
        "webhook",
        "cookie",
        "databaseurl",
        "connectionstring",
    ]
    .iter()
    .any(|needle| normalized == *needle || normalized.ends_with(needle))
}

fn is_large_private_key(key: &str) -> bool {
    let normalized = normalized_key(key);
    LARGE_PRIVATE_KEYS.contains(&normalized.as_str()) || normalized == "outputpath"
}

fn normalized_key(key: &str) -> String {
    key.chars()
        .filter(char::is_ascii_alphanumeric)
        .flat_map(char::to_lowercase)
        .collect()
}

fn sanitize_persisted_text(text: &str) -> String {
    redact_hashes(&redact_urls(&sanitize_private_text(text)))
}

fn sanitize_public_text(text: &str) -> String {
    sanitize_persisted_text(text)
}

fn redact_urls(text: &str) -> String {
    let bytes = text.as_bytes();
    let mut output = Vec::with_capacity(bytes.len());
    let mut copied_until = 0usize;
    let mut index = 0usize;
    while index < bytes.len() {
        let Some(mut end) = url_prefix_end(bytes, index) else {
            index += 1;
            continue;
        };
        let mut parenthesis_depth = 0usize;
        while end < bytes.len() {
            let byte = bytes[end];
            if byte.is_ascii_whitespace() || matches!(byte, b'"' | b'\'' | b'`' | b'<' | b'>') {
                break;
            }
            match byte {
                b'(' => parenthesis_depth += 1,
                b')' if parenthesis_depth == 0 => break,
                b')' => parenthesis_depth -= 1,
                _ => {}
            }
            end += 1;
        }
        output.extend_from_slice(&bytes[copied_until..index]);
        output.extend_from_slice(REDACTED.as_bytes());
        copied_until = end;
        index = end;
    }
    if copied_until == 0 {
        return text.to_string();
    }
    output.extend_from_slice(&bytes[copied_until..]);
    String::from_utf8(output).expect("URL redaction preserves UTF-8")
}

fn url_prefix_end(bytes: &[u8], start: usize) -> Option<usize> {
    if !bytes.get(start).is_some_and(u8::is_ascii_alphabetic)
        || start > 0
            && matches!(
                bytes[start - 1],
                b'a'..=b'z' | b'A'..=b'Z' | b'0'..=b'9' | b'+' | b'-' | b'.'
            )
    {
        return None;
    }
    let mut end = start + 1;
    while bytes
        .get(end)
        .is_some_and(|byte| byte.is_ascii_alphanumeric() || matches!(byte, b'+' | b'-' | b'.'))
    {
        end += 1;
    }
    bytes
        .get(end..end + 3)
        .is_some_and(|suffix| suffix == b"://")
        .then_some(end + 3)
}

fn redact_hashes(text: &str) -> String {
    let text = redact_labelled_digests(text);
    let text = text.as_str();
    let bytes = text.as_bytes();
    let mut output = Vec::with_capacity(bytes.len());
    let mut copied_until = 0usize;
    let mut index = 0usize;
    while index < bytes.len() {
        if !bytes[index].is_ascii_hexdigit() || index > 0 && bytes[index - 1].is_ascii_hexdigit() {
            index += 1;
            continue;
        }
        let mut end = index;
        while end < bytes.len() && bytes[end].is_ascii_hexdigit() {
            end += 1;
        }
        if end - index >= 40 {
            output.extend_from_slice(&bytes[copied_until..index]);
            output.extend_from_slice(REDACTED.as_bytes());
            copied_until = end;
        }
        index = end;
    }
    if copied_until == 0 {
        return text.to_string();
    }
    output.extend_from_slice(&bytes[copied_until..]);
    String::from_utf8(output).expect("hash redaction preserves UTF-8")
}

fn redact_labelled_digests(text: &str) -> String {
    let bytes = text.as_bytes();
    let mut output = Vec::with_capacity(bytes.len());
    let mut copied_until = 0usize;
    let mut index = 0usize;
    while index < bytes.len() {
        if index > 0 && bytes[index - 1].is_ascii_alphanumeric() {
            index += 1;
            continue;
        }
        let Some(label_len) = [b"sha512".as_slice(), b"sha256", b"sha1"]
            .into_iter()
            .find(|label| starts_ascii_case_insensitive(&bytes[index..], label))
            .map(|label| label.len())
        else {
            index += 1;
            continue;
        };
        let separator = index + label_len;
        if !bytes
            .get(separator)
            .is_some_and(|byte| matches!(byte, b':' | b'-'))
        {
            index += label_len;
            continue;
        }
        let token_start = separator + 1;
        let mut end = token_start;
        while bytes.get(end).is_some_and(|byte| {
            byte.is_ascii_alphanumeric() || matches!(byte, b'_' | b'+' | b'/' | b'=' | b':' | b'-')
        }) {
            end += 1;
        }
        if end - token_start < 20 {
            index = end.max(token_start);
            continue;
        }
        output.extend_from_slice(&bytes[copied_until..index]);
        output.extend_from_slice(REDACTED.as_bytes());
        copied_until = end;
        index = end;
    }
    if copied_until == 0 {
        return text.to_string();
    }
    output.extend_from_slice(&bytes[copied_until..]);
    String::from_utf8(output).expect("labelled digest redaction preserves UTF-8")
}

fn starts_ascii_case_insensitive(value: &[u8], prefix: &[u8]) -> bool {
    value
        .get(..prefix.len())
        .is_some_and(|candidate| candidate.eq_ignore_ascii_case(prefix))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn persistence_projection_rebuilds_a_safe_continuation_record() {
        let record = serde_json::json!({
            "id": "chat-1",
            "createdAt": 1,
            "updatedAt": 2,
            "title": "guide /Users/alice/private",
            "ownerId": "mallory",
            "privacyVersion": 999,
            "toolContext": {
                "mode": "instance",
                "root": "/Users/alice/Games",
                "instance": {
                    "root": "/Users/alice/Games",
                    "sourcePaths": ["/Users/alice/Games/pack"],
                    "modpackId": "pack",
                    "instanceId": "instance",
                    "mcVersion": "1.20.1",
                    "loader": "fabric"
                }
            },
            "messages": [{
                "id": "message-1",
                "role": "assistant",
                "parts": [
                    { "type": "reasoning", "text": "private chain" },
                    { "type": "text", "text": "https://storage.example/signed?token=secret 0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef" },
                    { "type": "tool-wiki_open", "toolCallId": "call-1", "output": {
                        "password": "secret", "content": "/Users/alice/private"
                    } }
                ]
            }]
        });
        let projected = project_conversation_record(&record, Some("alice")).unwrap();
        let serialized = serde_json::to_string(&projected).unwrap();
        assert!(!serialized.contains("private chain"));
        assert!(!serialized.contains("storage.example"));
        assert!(!serialized.contains("0123456789abcdef0123456789abcdef"));
        assert!(!serialized.contains("/Users/alice"));
        assert!(!serialized.contains("\"password\":\"secret\""));
        assert!(serialized.contains("toolCallId"));
        assert!(serialized.contains("instance_content_not_persisted"));
        assert_eq!(projected["ownerId"], "alice");
        assert_eq!(projected["privacyVersion"], 1);
        assert!(projected["toolContext"]["instance"].get("root").is_none());
    }

    #[test]
    fn public_projection_is_display_only_and_redacts_strings() {
        let raw = serde_json::json!({ "messages": [{
            "id": "message-1",
            "role": "assistant",
            "parts": [
                { "type": "reasoning", "text": "private" },
                { "type": "tool-wiki_open", "output": { "content": "secret" } },
                { "type": "text", "text": "See /Users/alice/private at https://example.com/a token=share-secret Bearer abc.def sk-abcdefghijklmnop sk_live_51ExampleLongCredential" },
                { "type": "tool-ask_user_question", "input": {
                    "question": "Choose /Users/alice/private",
                    "options": [{ "label": "A", "description": "https://secret.example" }]
                }, "output": { "selected": ["A"], "token": "drop" } }
            ]
        }] });
        let projected = project_public_share(&raw).unwrap();
        let serialized = serde_json::to_string(&projected).unwrap();
        assert!(serialized.contains("Choose"));
        assert!(!serialized.contains("message-1"));
        assert!(!serialized.contains("wiki_open"));
        assert!(!serialized.contains("/Users/alice"));
        assert!(!serialized.contains("example.com"));
        assert!(!serialized.contains("share-secret"));
        assert!(!serialized.contains("abc.def"));
        assert!(!serialized.contains("sk-abcdefghijklmnop"));
        assert!(!serialized.contains("sk_live_51ExampleLongCredential"));
        assert!(!serialized.contains("\"token\""));
    }

    #[test]
    fn persisted_text_redacts_generic_urls_and_arbitrarily_long_hashes() {
        let text = format!("s3://private-bucket/object {}", "a".repeat(160));
        let sanitized = sanitize_persisted_text(&text);
        assert!(!sanitized.contains("private-bucket"));
        assert!(!sanitized.contains(&"a".repeat(160)));
    }

    #[test]
    fn conversation_text_redacts_parenthesized_urls_and_labelled_digests() {
        let text = concat!(
            "See https://internal.example/report(user)?sig=private-value next; ",
            "sha256:QWxhZGRpbjpvcGVuIHNlc2FtZV9wcml2YXRlLXNpZw== ",
            "sha512-AbCdEfGhIjKlMnOpQrStUvWxYz0123456789_-private ",
            "gameplay=create:crushing_wheel ",
            "unlabelled=QWxhZGRpbjpvcGVuIHNlc2FtZQ==",
        );

        let sanitized = sanitize_conversation_text(text);
        assert!(!sanitized.contains("internal.example"));
        assert!(!sanitized.contains("user)?sig=private-value"));
        assert!(!sanitized.contains("QWxhZGRpbjpvcGVuIHNlc2FtZV9wcml2YXRlLXNpZw"));
        assert!(!sanitized.contains("AbCdEfGhIjKlMnOpQrStUvWxYz0123456789_-private"));
        assert!(sanitized.contains("next"));
        assert!(sanitized.contains("create:crushing_wheel"));
        assert!(sanitized.contains("QWxhZGRpbjpvcGVuIHNlc2FtZQ=="));
    }

    #[test]
    fn persistence_projection_sanitizes_object_keys_without_dropping_collisions() {
        let mut metadata = Map::new();
        metadata.insert("https://one.example/private".into(), "first".into());
        metadata.insert("https://two.example/private".into(), "second".into());
        metadata.insert("a".repeat(40), "third".into());
        metadata.insert("b".repeat(40), "fourth".into());
        let record = serde_json::json!({
            "id": "chat-keys",
            "messages": [{
                "role": "assistant",
                "parts": [{ "type": "text", "text": "safe" }],
                "metadata": metadata,
            }],
        });

        let projected = project_conversation_record(&record, None).unwrap();
        let metadata = projected["messages"][0]["metadata"].as_object().unwrap();
        let serialized = serde_json::to_string(metadata).unwrap();
        assert_eq!(metadata.len(), 4);
        for key in [REDACTED, "[REDACTED]#2", "[REDACTED]#3", "[REDACTED]#4"] {
            assert!(metadata.contains_key(key));
        }
        let mut values = metadata
            .values()
            .filter_map(Value::as_str)
            .collect::<Vec<_>>();
        values.sort_unstable();
        assert_eq!(values, ["first", "fourth", "second", "third"]);
        assert!(!serialized.contains("one.example"));
        assert!(!serialized.contains("two.example"));
        assert!(!serialized.contains(&"a".repeat(40)));
        assert!(!serialized.contains(&"b".repeat(40)));
    }

    #[test]
    fn visibility_never_exposes_another_accounts_records() {
        let anonymous = serde_json::json!({ "ownerId": null });
        let alice = serde_json::json!({ "ownerId": "alice" });
        assert!(conversation_record_visible_to(&anonymous, Some("bob")));
        assert!(!conversation_record_visible_to(&alice, Some("bob")));
        assert!(conversation_record_visible_to(&alice, Some("alice")));
        assert!(!conversation_record_visible_to(&alice, None));
    }

    #[test]
    fn sensitive_key_matching_does_not_hide_author_metadata() {
        let record = serde_json::json!({
            "id": "chat-author",
            "messages": [{
                "role": "assistant",
                "author": "Guide Writer",
                "apiToken": "private-token",
                "parts": [{ "type": "text", "text": "safe" }],
            }],
        });

        let projected = project_conversation_record(&record, None).unwrap();
        assert_eq!(projected["messages"][0]["author"], "Guide Writer");
        assert_eq!(projected["messages"][0]["apiToken"], REDACTED);
    }
}
